diff --git a/CMakeLists.txt b/CMakeLists.txt
index 7415eab..a5b8027 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -561,7 +561,7 @@ install (TARGETS glog
   ARCHIVE DESTINATION lib)
 
 if (gflags_FOUND)
-  set (gflags_DEPENDENCY "find_dependency (gflags ${gflags_VERSION})")
+  set (gflags_DEPENDENCY "find_package(GFLAGS ${gflags_VERSION})")
 endif (gflags_FOUND)
 
 configure_package_config_file (glog-config.cmake.in
