diff --git a/caca0.h b/caca0.h
index 608f483..7e2067f 100644
--- a/caca0.h
+++ b/caca0.h
@@ -145,7 +145,7 @@ enum caca_feature
 #define caca_get_color_name __caca0_get_color_name
 #define caca_putchar(x, y, c) caca_put_char(__caca0_cv, x, y, c)
 #define caca_putstr(x, y, s) caca_put_str(__caca0_cv, x, y, s)
-#define caca_printf(x, y, f, z...) caca_printf(__caca0_cv, x, y, f, ##z)
+#define caca_printf(x, y, f, __VA_ARGS__) caca_printf(__caca0_cv, x, y, f, ##__VA_ARGS__)
 #define caca_clear() caca_clear_canvas(__caca0_cv)
 
 #define caca_draw_line(x, y, z, t, c) \
