diff --git a/CMakeLists.txt b/CMakeLists.txt
index d49c31e..09152bd 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -5,6 +5,8 @@
 cmake_minimum_required(VERSION 3.9)
 project(leveldb VERSION 1.21.0 LANGUAGES C CXX)
 
+set(PLATFORM_CONSTANT LEVELDB_PLATFORM_WINDOWS=1)
+
 # This project can use C11, but will gracefully decay down to C89.
 set(CMAKE_C_STANDARD 11)
 set(CMAKE_C_STANDARD_REQUIRED OFF)
@@ -74,10 +76,10 @@ include_directories(
   "${PROJECT_SOURCE_DIR}"
 )
 
-if(BUILD_SHARED_LIBS)
+if(BUILD_SHARED_LIBS AND UNIX)
   # Only export LEVELDB_EXPORT symbols from the shared library.
   add_compile_options(-fvisibility=hidden)
-endif(BUILD_SHARED_LIBS)
+endif(BUILD_SHARED_LIBS AND UNIX)
 
 add_library(leveldb "")
 target_sources(leveldb
@@ -172,12 +174,24 @@ target_sources(leveldb
     "${LEVELDB_PUBLIC_INCLUDE_DIR}/write_batch.h"
 )
 
-# POSIX code is specified separately so we can leave it out in the future.
-target_sources(leveldb
-  PRIVATE
-    "${PROJECT_SOURCE_DIR}/util/env_posix.cc"
-    "${PROJECT_SOURCE_DIR}/util/posix_logger.h"
-)
+if(UNIX)
+  # POSIX code is specified separately so we can leave it out in the future.
+  target_sources(leveldb
+    PRIVATE
+      "${PROJECT_SOURCE_DIR}/util/env_posix.cc"
+      "${PROJECT_SOURCE_DIR}/util/posix_logger.h"
+  )
+elseif(WIN32)
+  target_sources(leveldb
+    PRIVATE
+      "${PROJECT_SOURCE_DIR}/port/port_win.h"
+      "${PROJECT_SOURCE_DIR}/port/port_win.cc"
+      "${PROJECT_SOURCE_DIR}/util/env_win.cc"
+  )
+  target_link_libraries(leveldb iphlpapi shlwapi)
+endif()
+
+
 
 # MemEnv is not part of the interface and could be pulled to a separate library.
 target_sources(leveldb
@@ -196,7 +210,7 @@ target_compile_definitions(leveldb
     # Used by include/export.h when building shared libraries.
     LEVELDB_COMPILE_LIBRARY
     # Used by port/port.h.
-    LEVELDB_PLATFORM_POSIX=1
+    ${PLATFORM_CONSTANT}
 )
 if (NOT HAVE_CXX17_HAS_INCLUDE)
   target_compile_definitions(leveldb
@@ -258,8 +272,15 @@ if(LEVELDB_BUILD_TESTS)
     target_link_libraries("${test_target_name}" leveldb)
     target_compile_definitions("${test_target_name}"
       PRIVATE
-        LEVELDB_PLATFORM_POSIX=1
+        ${PLATFORM_CONSTANT}
     )
+    if(WIN32)
+      target_sources("${test_target_name}"
+        PRIVATE
+          "${PROJECT_SOURCE_DIR}/port/port_win.h"
+          "${PROJECT_SOURCE_DIR}/port/port_win.cc"
+      )
+    endif()
     if (NOT HAVE_CXX17_HAS_INCLUDE)
       target_compile_definitions("${test_target_name}"
         PRIVATE
@@ -307,7 +328,9 @@ if(LEVELDB_BUILD_TESTS)
 
     # TODO(costan): This test also uses
     #               "${PROJECT_SOURCE_DIR}/util/env_posix_test_helper.h"
-    leveldb_test("${PROJECT_SOURCE_DIR}/util/env_posix_test.cc")
+    if(UNIX)
+      leveldb_test("${PROJECT_SOURCE_DIR}/util/env_posix_test.cc")
+    endif(UNIX)
   endif(NOT BUILD_SHARED_LIBS)
 endif(LEVELDB_BUILD_TESTS)
 
@@ -331,7 +354,7 @@ if(LEVELDB_BUILD_BENCHMARKS)
     target_link_libraries("${bench_target_name}" leveldb)
     target_compile_definitions("${bench_target_name}"
       PRIVATE
-        LEVELDB_PLATFORM_POSIX=1
+        ${PLATFORM_CONSTANT}
     )
     if (NOT HAVE_CXX17_HAS_INCLUDE)
       target_compile_definitions("${bench_target_name}"
@@ -408,12 +431,17 @@ if(LEVELDB_INSTALL)
   install(
     EXPORT leveldbTargets
     NAMESPACE leveldb::
-    DESTINATION "${CMAKE_INSTALL_LIBDIR}/cmake/leveldb"
+    DESTINATION "${CMAKE_INSTALL_DATADIR}"
   )
   install(
     FILES
       "${PROJECT_SOURCE_DIR}/cmake/leveldbConfig.cmake"
       "${PROJECT_BINARY_DIR}/leveldbConfigVersion.cmake"
-    DESTINATION "${CMAKE_INSTALL_LIBDIR}/cmake/leveldb"
+    DESTINATION "${CMAKE_INSTALL_DATADIR}"
   )
+  install(TARGETS leveldbutil
+    RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR}
+    LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
+    ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR}
+)
 endif(LEVELDB_INSTALL)
diff --git a/port/port.h b/port/port.h
index 0975fed..3cf259b 100644
--- a/port/port.h
+++ b/port/port.h
@@ -14,6 +14,8 @@
 #  include "port/port_stdcxx.h"
 #elif defined(LEVELDB_PLATFORM_CHROMIUM)
 #  include "port/port_chromium.h"
+#elif defined(LEVELDB_PLATFORM_WINDOWS)
+#  include "port/port_win.h"
 #endif
 
 #endif  // STORAGE_LEVELDB_PORT_PORT_H_
